cupy/cuda/cupy_thrust.cu

```diff
@@ -129,3 +132,40 @@ template void cupy::thrust::_argsort<cpy_long>(size_t *, void *, size_t);
 template void cupy::thrust::_argsort<cpy_ulong>(size_t *, void *, size_t);
 template void cupy::thrust::_argsort<cpy_float>(size_t *, void *, size_t);
 template void cupy::thrust::_argsort<cpy_double>(size_t *, void *, size_t);
+
+
+/*
+ * shuffle
+ */
+
+struct Shuffle
+{
+    uint32_t seed;
+    Shuffle(uint32_t s) { seed = s; }
+    __device__ uint32_t operator()(uint32_t kn) const
+    {
+        minstd_rand randEngine(seed);
+        randEngine.discard(kn);
+        return randEngine();
+    }
+};
+
+template <typename T>
+void cupy::thrust::_shuffle(void *start, ptrdiff_t num, uint32_t seed) {
+    device_ptr<T> dp_first = device_pointer_cast((T *)start);
+    device_ptr<T> dp_last  = device_pointer_cast((T *)start + num);
+    device_vector<uint16_t> d_keys(num);
```

Comment:
Because I don't have any particular intension to `uint16_t`, I've changed it to `uint32_t`.